src/shader/shaderGGXFunctions.js

```diff
@@ -72,17 +72,18 @@ float ggxDistribution( vec3 halfVector, float roughness ) {
 
 	// See equation (33) from reference [0]
 	float a2 = roughness * roughness;
+	a2 = max( EPSILON, a2 );
 	float cosTheta = halfVector.z;
 	float cosTheta4 = pow( cosTheta, 4.0 );
 
 	if ( cosTheta == 0.0 ) return 0.0;
 
-	float theta = acosApprox( halfVector.z );
+	float theta = acosSafe( halfVector.z );
 	float tanTheta = tan( theta );
 	float tanTheta2 = pow( tanTheta, 2.0 );
 
 	float denom = PI * cosTheta4 * pow( a2 + tanTheta2, 2.0 );
-	return denom < 1e-6 ? 1.0 : a2 / denom;
+	return ( a2 / denom );
```

Comment:
With this change (and `max( EPSILON, a2 )`) I'm still seeing halos:

![image](https://user-images.githubusercontent.com/734200/173253947-0ad06e9d-246f-47c0-9877-9b92591a6511.png) ![image](https://user-images.githubusercontent.com/734200/173254097-a2de3147-ebca-4cec-8c6a-43b0894a52bc.png)

This line you had here previously seemed to behave better, though I'm not sure I follow exactly where it came from:

```
return ( denom < 1e-12 ) ? PI : ( a2 / denom );
```

Reply:
Try removing `envColor` from: 

`gl_FragColor.rgb += attenuatedColor * environmentIntensity * envColor * throughputColor * sampleColor * misWeight / envPdf;`

in `PhysicalPathTracingMaterial.js`

This seems to solve the halo issue but I wanted to examine more closely what is causing a correct looking line to be a factor here, perhaps some quantity is being included twice. 

The above is in an MIS-only `#ifdef` and with MIS-off the halos are still present, this might be informative.

I think `max( EPSILON, a2 )` and `acosSafe` are more sensible changes here though and this is revealing an issue elsewhere, I'll try to track it down.

Reply:
I'll merge this for now since I do think it's a big improvement - I made #179 to track any other thoughts on what might be causing this. Thanks again!